src/appleseed/renderer/modeling/project/assethandler.cpp

```diff
@@ -0,0 +1,256 @@
+#include "renderer/modeling/project/renderlayerrule.h"
+#include "renderer/modeling/scene/scene.h"
+
+// appleseed.foundation headers.
+#include "foundation/platform/path.h"
+#include "foundation/utility/containers/specializedarrays.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/searchpaths.h"
+#include "foundation/utility/string.h"
+
+// Boost headers.
+#include "boost/filesystem/operations.hpp"
+
+// Standard headers.
+#include <cstddef>
+#include <vector>
+
+using namespace boost;
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+//
+// AssetHandler class implementation.
+//
+
+namespace
+{
+    string convert_to_posix(string path)
+    {
+        replace(path.begin(), path.end(), '\\', '/');
+        return path;
+    }
+
+    bool safe_copy_file(
+        const filesystem::path& source_path,
+        const filesystem::path& dest_path)
+    {
+        try
+        {
+            filesystem::create_directories(dest_path.parent_path());
+            filesystem::copy_file(source_path, dest_path, filesystem::copy_option::overwrite_if_exists);
+            return true;
+        }
+        catch (const std::exception& e)
+        {
+            RENDERER_LOG_ERROR(
+                "failed to copy %s to %s: %s.",
+                source_path.string().c_str(),
+                dest_path.string().c_str(),
+                e.what());
+            return false;
+        }
+    }
+
+    template <typename EntityCollection>
+    void do_collect_asset_paths(
+        StringArray&            paths,
+        const EntityCollection& entities)
+    {
+        for (const_each<EntityCollection> i = entities; i; ++i)
+            i->collect_asset_paths(paths);
+    }
+
+    template <typename EntityCollection>
+    void do_update_asset_paths(
+        const StringDictionary& mappings,
+        EntityCollection&       entities)
+    {
+        for (each<EntityCollection> i = entities; i; ++i)
+            i->update_asset_paths(mappings);
+    }
+}
+
+AssetHandler::AssetHandler(
+    const Project&      project,
+    const char*         filepath,
+    const Mode          mode)
+  : m_project(project)
+  , m_project_old_root_path(filesystem::absolute(project.get_path()))
+  , m_project_new_root_path(filesystem::absolute(filepath))
+  , m_project_old_root_dir(m_project_old_root_path.parent_path())
+  , m_project_new_root_dir(m_project_new_root_path.parent_path())
+  , m_mode(mode)
+{
+}
+
+bool AssetHandler::handle_assets() const
+{
+    StringArray paths;
+
+    Scene* scene = m_project.get_scene();
+    if (scene)
+        scene->collect_asset_paths(paths);
+
+    Frame* frame = m_project.get_frame();
+    if (frame)
+        frame->collect_asset_paths(paths);
+
+    do_collect_asset_paths(paths, m_project.render_layer_rules());
+    do_collect_asset_paths(paths, m_project.configurations());
+
+    vector<string> unique_paths = array_vector<vector<string> >(paths);
```

Comment:
> the parser would have been confused by the presence of a right-shift token.

I'm very surprised about this. I have the impression I never had this problem. Maybe because `array_vector` and `vector` you use are `typedef` or macro or things like that? Anyway, I lean something.